ui/AGENTS.md

```diff
@@ -0,0 +1,207 @@
+npm run build             # Build for production
+npm run clean             # Clean artifacts and test results
+npm run start:ocp-plugin  # OpenShift Console plugin dev server
+npm run build:ocp-plugin  # Build OpenShift Console plugin
+```
+
+### Testing
+
+```bash
+npm run test                                  # Unit tests (Vitest)
+npm run test-coverage                         # Unit tests with coverage
+npm run test -- --testNamePattern="Name"      # Single unit test
+npm run test -- src/path/to/test.test.ts      # Test by file path
+npm run test-e2e                              # E2E tests (Cypress)
+npm run cypress-spec -- "test-name.js"        # Single E2E test
+npm run test-component                        # Component tests
+npm run cypress-open                          # Interactive test runner
+```
+
+### Code Quality
+
+```bash
+npm run lint:fast-dev      # Quick ESLint check (skips slow rules)
+npm run lint:fast-dev:fix  # Quick ESLint with auto-fix
+npm run lint:fix           # Full ESLint with auto-fix
+npm run tsc                # TypeScript type checking
+```
+
+**Important:** Use IDE diagnostics (real-time linting) instead of running full lint commands — they're much faster for development. Use `lint:fast-dev` only if you need to run lint from the command line.
+
+---
+
+## Code Guidelines
+
+### Coding Patterns
+
+#### Conditional Rendering
+
+Process data before JSX to avoid unnecessary renders:
+
+```tsx
+// Process first, then render
+const validItems = rawData ? normalizeArray(rawData).filter(isValid) : [];
+return validItems.length > 0 ? (
+  <Stack>
+    {validItems.map((item) => (
+      <Component key={item} />
+    ))}
+  </Stack>
+) : (
+  'Fallback'
+);
+```
+
+### Data Fetching
+
+Use custom hooks for data fetching with built-in state management:
+
+```typescript
+// Queries (GET)
+const { data, isLoading, error, refetch } = useRestQuery(() =>
+  fetchMyResources(filters, page)
+);
+
+// Mutations (POST/PUT/DELETE)
+const { mutate, isLoading, isSuccess } = useRestMutation(createMyResource, {
+  onSuccess: () => refetch(),
+});
+```
+
+**Service pattern:** Create service functions in `src/services/` that extract and return data
+
+```typescript
+export function fetchMyResource(id: string): Promise<MyResource> {
+  return axios
+    .get<MyResource>(`/v1/resources/${id}`)
+    .then((resource) => resource.data);
+}
+```
+
+**Best practices:**
+
+- Always check existing services before implementing new endpoints
+- Use cancellation for heavy queries to prevent memory leaks
+- Avoid creating wrapper hooks around `useRestQuery`—pass fetch functions directly
+
+### State Management
+
+- **Prefer React built-ins:** `useState`, `useReducer`, and React Context for new code
+- **Avoid Redux/Redux-Saga:** Only modify existing code when migrating away
+- **Component state:** Use `useState` for component-specific data
+- **Shared state:** Use React Context for cross-component state
```

Comment:
I'm _slightly_ concerned that 50% of the state management bullets mention React Context, when that pattern is very much the exception for state management in our code.

Reply:
Hmm, maybe a flow diagram for how to think about state management would help alleviate the worries. I felt like it helped in deciding which custom wrapper to use for Selects. We can also emphasize using component-level state.